scipy/optimize/_linprog.py

```diff
@@ -135,742 +141,1009 @@ def linprog_terse_callback(xk, **kwargs):
-             2 : Problem appears to be infeasible
-             3 : Problem appears to be unbounded
+    """
 
-        nit : int
-            The number of iterations performed.
-        message : str
-            A string descriptor of the exit status of the optimization.
+    if sps.issparse(A_eq):
+        sparse = True
+        A_eq = sps.lil_matrix(A_eq)
+        A_ub = sps.lil_matrix(A_ub)
 
-    Examples
-    --------
-    Consider the following problem:
+        def hstack(blocks):
+            return sps.hstack(blocks, format="lil")
 
-    Minimize: f = -1*x[0] + 4*x[1]
+        def vstack(blocks):
+            return sps.vstack(blocks, format="lil")
 
-    Subject to: -3*x[0] + 1*x[1] <= 6
-                 1*x[0] + 2*x[1] <= 4
-                            x[1] >= -3
+        zeros = sps.lil_matrix
+        eye = sps.eye
+    else:
+        sparse = False
+        hstack = np.hstack
+        vstack = np.vstack
+        zeros = np.zeros
+        eye = np.eye
+
+    fixed_x = set()
+    if len(undo) > 0:
+        # these are indices of variables removed from the problem
+        # however, their bounds are still part of the bounds list
+        fixed_x = set(undo[0])
+    # they are needed elsewhere, but not here
+    bounds = [bounds[i] for i in range(len(bounds)) if i not in fixed_x]
+    # in retrospect, the standard form of bounds should have been an n x 2
+    # array. maybe change it someday.
+
+    # modify problem such that all variables have only non-negativity bounds
+
+    bounds = np.array(bounds)
+    lbs = bounds[:, 0]
+    ubs = bounds[:, 1]
+    m_ub, n_ub = A_ub.shape
+
+    lb_none = np.equal(lbs, None)
+    ub_none = np.equal(ubs, None)
+    lb_some = np.logical_not(lb_none)
+    ub_some = np.logical_not(ub_none)
+
+    # if preprocessing is on, lb == ub can't happen
+    # if preprocessing is off, then it would be best to convert that
+    # to an equality constraint, but it's tricky to make the other
+    # required modifications from inside here.
+
+    # unbounded below: substitute xi = -xi' (unbounded above)
+    l_nolb_someub = np.logical_and(lb_none, ub_some)
+    i_nolb = np.nonzero(l_nolb_someub)[0]
+    lbs[l_nolb_someub], ubs[l_nolb_someub] = (
+        -ubs[l_nolb_someub], lbs[l_nolb_someub])
+    lb_none = np.equal(lbs, None)
+    ub_none = np.equal(ubs, None)
+    lb_some = np.logical_not(lb_none)
+    ub_some = np.logical_not(ub_none)
+    c[i_nolb] *= -1
+    if len(i_nolb) > 0:
+        if A_ub.shape[0] > 0:  # sometimes needed for sparse arrays... weird
+            A_ub[:, i_nolb] *= -1
+        if A_eq.shape[0] > 0:
+            A_eq[:, i_nolb] *= -1
+
+    # upper bound: add inequality constraint
+    i_newub = np.nonzero(ub_some)[0]
+    ub_newub = ubs[ub_some]
+    n_bounds = np.count_nonzero(ub_some)
+    A_ub = vstack((A_ub, zeros((n_bounds, A_ub.shape[1]))))
+    b_ub = np.concatenate((b_ub, np.zeros(n_bounds)))
+    A_ub[range(m_ub, A_ub.shape[0]), i_newub] = 1
```

Comment:
```
A_ub[m_ub:, i_newub] = 1.
```

Reply:
Agreed. Weird considering the next line...